File: src/api/db/models/llm_provider.py
"""LLM Provider database model.

Stores configuration for different LLM providers (Anthropic, OpenAI, Google, Custom).
API keys are stored encrypted using Fernet encryption.
"""

import uuid
from sqlalchemy import JSON, Boolean, Column, String, DateTime, ForeignKey, func, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from ..base import Base


class LlmProvider(Base):
    """Model for storing LLM provider configurations.

    Attributes:
        id: Unique identifier for the provider.
        name: User-friendly name (e.g., "Production Claude").
        provider_type: Type of provider (anthropic, openai, google, custom).
        base_url: Optional base URL for proxies or custom endpoints.
        api_key_encrypted: Fernet-encrypted API key.
        config: Provider-specific configuration as JSON.
        models: List of available model identifiers as JSON array.
        is_active: Whether the provider is enabled.
        is_default: Whether this is the default provider (only one can be default).
        last_health_check_at: Timestamp of last health check.
        last_health_check_status: Result of last health check (success/error).
        last_health_check_error: Error message if last health check failed.
        created_by: User who created this provider.
        updated_by: User who last updated this provider.
        created_at: Creation timestamp.
        updated_at: Last update timestamp.
    """

    __tablename__ = "llm_providers"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    # Provider identification
    name = Column(String(100), nullable=False)
    provider_type = Column(String(50), nullable=False)  # anthropic, openai, google, custom

    # Connection configuration
    base_url = Column(String(500), nullable=True)
    api_key_encrypted = Column(Text, nullable=True)

    # Provider-specific config (JSON)
    # For anthropic/openai: {"organization_id": "..."}
    # For custom: {"auth_type": "bearer|basic|api_key_header|none", "auth_header_name": "...", "custom_headers": {...}}
    config = Column(JSON, nullable=False, default=dict)

    # Available models for this provider (JSON array)
    models = Column(JSON, nullable=False, default=list)

    # Status flags
    is_active = Column(Boolean, default=True)
    is_default = Column(Boolean, default=False)

    # Health check tracking
    last_health_check_at = Column(DateTime(timezone=True), nullable=True)
    last_health_check_status = Column(String(20), nullable=True)  # success, error
    last_health_check_error = Column(Text, nullable=True)

    # Audit fields
    created_by = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )
    updated_by = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    # Relationships
    creator = relationship("User", foreign_keys=[created_by])
    updater = relationship("User", foreign_keys=[updated_by])

    def __repr__(self):
        return f"<LlmProvider(id={self.id}, name='{self.name}', type='{self.provider_type}')>"

File: frontend/src/types/LlmProvider.ts
/**
 * LLM Provider type definitions for the frontend.
 */

/**
 * Supported LLM provider types.
 */
export type ProviderType = 'anthropic' | 'openai' | 'google' | 'custom';

/**
 * Health check status values.
 */
export type HealthStatus = 'success' | 'error' | null;

/**
 * LLM Provider configuration stored in the database.
 */
export interface LlmProvider {
  id: string;
  name: string;
  provider_type: ProviderType;
  base_url: string | null;
  has_api_key: boolean;
  config: ProviderConfig;
  models: string[];
  is_active: boolean;
  is_default: boolean;
  last_health_check_at: string | null;
  last_health_check_status: HealthStatus;
  last_health_check_error: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * Provider-specific configuration options.
 */
export interface ProviderConfig {
  // Authentication type for custom providers
  auth_type?: 'none' | 'bearer' | 'api_key_header' | 'basic';
  // Custom header name for API key authentication
  auth_header_name?: string;
  // Organization ID for OpenAI
  organization_id?: string;
  // Whether this is an OpenAI-compatible API (vs Ollama-compatible)
  openai_compatible?: boolean;
  // Auto-discover models from the API
  auto_discover_models?: boolean;
  // Custom headers to send with requests
  custom_headers?: Record<string, string>;
}

/**
 * Payload for creating a new LLM provider.
 */
export interface LlmProviderCreate {
  name: string;
  provider_type: ProviderType;
  base_url?: string | null;
  api_key?: string | null;
  config?: ProviderConfig;
  models?: string[];
  is_active?: boolean;
  is_default?: boolean;
}

/**
 * Payload for updating an LLM provider.
 */
export interface LlmProviderUpdate {
  name?: string;
  base_url?: string | null;
  api_key?: string | null;
  config?: ProviderConfig;
  models?: string[];
  is_active?: boolean;
  is_default?: boolean;
}

/**
 * Response from provider list endpoint.
 */
export interface LlmProviderListResponse {
  providers: LlmProvider[];
  total: number;
}

/**
 * Health check result from test connection endpoint.
 */
export interface HealthCheckResult {
  success: boolean;
  provider_id: string;
  status: 'success' | 'error';
  message?: string;
  response_time_ms?: number;
}

/**
 * Available model for selection.
 */
export interface AvailableModel {
  provider_id: string;
  provider_name: string;
  provider_type: string;
  model_id: string;
  display_name: string;
  is_default_provider: boolean;
}

/**
 * Response from available models endpoint.
 */
export interface AvailableModelsResponse {
  models: AvailableModel[];
}

/**
 * Display labels for provider types.
 */
export const PROVIDER_LABELS: Record<ProviderType, string> = {
  anthropic: 'Anthropic (Claude)',
  openai: 'OpenAI (GPT)',
  google: 'Google (Gemini)',
  custom: 'Custom Provider',
};

/**
 * Provider type colors for badges.
 */
export const PROVIDER_COLORS: Record<ProviderType, string> = {
  anthropic: 'bg-orange-100 text-orange-800',
  openai: 'bg-green-100 text-green-800',
  google: 'bg-blue-100 text-blue-800',
  custom: 'bg-purple-100 text-purple-800',
};

/**
 * Form field configuration for each provider type.
 */
export interface ProviderFieldConfig {
  showApiKey: boolean;
  apiKeyRequired: boolean;
  showBaseUrl: boolean;
  baseUrlRequired: boolean;
  showAuthType: boolean;
  showOrganizationId: boolean;
  baseUrlPlaceholder: string;
  apiKeyPlaceholder: string;
}

export const PROVIDER_FIELDS: Record<ProviderType, ProviderFieldConfig> = {
  anthropic: {
    showApiKey: true,
    apiKeyRequired: true,
    showBaseUrl: true,
    baseUrlRequired: false,
    showAuthType: false,
    showOrganizationId: false,
    baseUrlPlaceholder: 'https://api.anthropic.com (optional, for proxies)',
    apiKeyPlaceholder: 'sk-ant-...',
  },
  openai: {
    showApiKey: true,
    apiKeyRequired: true,
    showBaseUrl: true,
    baseUrlRequired: false,
    showAuthType: false,
    showOrganizationId: true,
    baseUrlPlaceholder: 'https://api.openai.com (optional, for proxies)',
    apiKeyPlaceholder: 'sk-...',
  },
  google: {
    showApiKey: true,
    apiKeyRequired: true,
    showBaseUrl: false,
    baseUrlRequired: false,
    showAuthType: false,
    showOrganizationId: false,
    baseUrlPlaceholder: '',
    apiKeyPlaceholder: 'AI...',
  },
  custom: {
    showApiKey: true,
    apiKeyRequired: false,
    showBaseUrl: true,
    baseUrlRequired: true,
    showAuthType: true,
    showOrganizationId: false,
    baseUrlPlaceholder: 'https://your-llm-server.com or http://localhost:11434',
    apiKeyPlaceholder: 'API key (if required)',
  },
};

/**
 * Authentication type options for custom providers.
 */
export const AUTH_TYPE_OPTIONS = [
  { value: 'none', label: 'No Authentication' },
  { value: 'bearer', label: 'Bearer Token' },
  { value: 'api_key_header', label: 'API Key Header' },
  { value: 'basic', label: 'Basic Auth' },
];
